ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala

```diff
@@ -290,28 +317,38 @@ class WebSocketService(
     Flow[Message]
       .mapAsync(1) {
         case msg: TextMessage =>
-          msg.toStrict(config.maxDuration).map(m => Q.parse(decoder, m.text))
+          msg.toStrict(config.maxDuration).map { m =>
+            SprayJson.parse(m.text).liftErr(InvalidUserInput)
+          }
         case _ =>
           Future successful -\/(
             InvalidUserInput("Cannot process your input, Expect a single JSON message"))
       }
+      .via(withOptPrefix { ejv: (InvalidUserInput \/ JsValue) =>
+        ejv
+          .flatMap(jv =>
+            Bitraverse[\/].bisequence(
+              readStartingOffset(jv) toLeftDisjunction Q.parse(decoder, jv).strengthL(None)))
+          .sequence: domain.StartingOffset \/ (Error \/ (Option[domain.StartingOffset], A))
+      }((offPrefix, ejv) => ejv flatMap (jv => Q.parse(decoder, jv) strengthL Some(offPrefix))))
```

Comment:
this entire block + `readStartingOffset` could have been simplified with derived JSON Protofol for `domain.StreamingRequest` variant. Hard to comprehend what is going on here.

Reply:
You might be right about `readStartingOffset`, but that would do nothing to simplify this block.

Reply:
The reason I didn't derive is that I do not want to treat all JSON parse errors equally;

1. a missing ``offset`` key should mean skip to the required body,
2. a present ``offset`` key with non-string should fail, and
3. mixing an ``offset`` key with any other key should fail,

but when you derive parsers

```
final case class Foo(foo: Int)
"""{"foo":1, "offset":"hi"}""".parseJson.convertTo[Foo]
// ⇑ evaluates to Foo(1)
```

Once you add all this special handling, bothering with the derived parser becomes less a convenience and more a...bother. I can still work it in if you'd prefer, but it's not going to be as pretty as you might be imagining. Let me know what you think.

I'll also think about what I can do with the definition of `withOptPrefix`, which I believe can be made more complicated to simplify the call to it that you've directly highlighted.

Reply:
See 9abd4e8cd8 for a `withOptPrefix` redesign.